File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    private static final int[][] DIRECTIONS = {
        {-1, 0}, {1, 0}, {0, -1}, {0, 1}
    };

    public int numIslands(char[][] grid) {
        if(grid == null || grid.length == 0) return 0;

        int res = 0;

        for(int i=0; i<grid.length; i++){
            for(int j=0; j<grid[0].length; j++){
                if(grid[i][j] == '1'){
                    res++;
                    dfs(grid, i, j);
                }
            }
        }

        return res;
    }

    private void dfs(char[][] grid, int x, int y){
        if(x<0 || x>=grid.length || y<0 || y>=grid[0].length || grid[x][y] == '0'){
                return;
        }

        grid[x][y] = '0';

        for(int[] direction : DIRECTIONS){
            int newX = x + direction[0];
            int newY = y + direction[1];
            dfs(grid, newX, newY);
        }
    }
}

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        m = len(grid)
        n = len(grid[0])

        def dfs(i, j):
            if i<0 or i>=m or j<0 or j>=n or grid[i][j]=='0':
                return

            grid[i][j] = '0'

            for dx, dy in [(-1,0), (1,0), (0,-1), (0,1)]:
                dfs(i + dx, j + dy)

        count = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    count+=1
                    dfs(i,j)

        return count
